### ISSUE TITLE
HoistField Cleanups

### ISSUE THREAD

**Issue by amcclain** [exhi/hoist-react#475]
I had spent a while a few weeks ago trying to clean up HoistFields and their docs, but then I didn’t get to complete that and I think the branch I openeed is probably pretty dead now in terms of being usable for merging. But that said, unless something has changed there is probably still some good cleanup I would like us to do.

I made two commits on that branch - https://github.com/exhi/hoist-react/commits/fieldCodeCleanups - I think they show pretty much what I was trying to accomplish in terms of docs / comments / code style.  

Basically when I started really looking at those classes there were a bunch of inconsistencies with what was documented and how.

A bigger change I would like to consider was removing the `delegateProps` notion from HoistField, as this was rather opaque, made the propType-based docs incomplete, and wasn't being used consistently.

**Comment by lbwexler** [comment]
Let's go ahead and remove the delegate props and see how things work without it? It was just an experiment of mine and always intended to just be an implementation helper for authors of Hoist Fields themselves, and never something intended to be visible to Application.

Was mainly trying to DRY up things in the implementation.

**Comment by amcclain** [comment]
As part of this, would also like to **document / formalize the requirements and handling of record IDs**. See discussion within #488, including potentially problematic code that currently generates missing IDs using the index of the raw data object within the records collection: 

https://github.com/exhi/hoist-react/blob/0dc739d2ddeeff0ba5aa8235afd9ee07e57bb085/data/LocalStore.js#L114

**Comment by amcclain** [comment]
Grabbing this ticket back as part of my ongong docs work.

**Comment by lbwexler** [comment]
I have a pretty major refactor here in progress that flattens out this whole hierarchy, and simplifies the documentation.  Also removes the much beloved delegateProps.

**Comment by lbwexler** [comment]
Closing this -- Anselm wrangled these changes into the product earlier this fall.

Created the following to capture the one somewhat unrelated item that was not dealt with here:
https://github.com/exhi/hoist-react/issues/844
